tests/test_retokenizer.py

```diff
@@ -0,0 +1,80 @@
+import random
+import unittest
+from opustrainer.modifiers.retokenize import RetokenizeModifier
+
+
+VOCAB = 'contrib/test-data/vocab.zhen.spm'
+
+
+class TestTokenizer(unittest.TestCase):
+  maxDiff = None
+
+  def test_identity(self):
+    """The default tokenizer and detokenizer together should have no impact on
+    the output."""
+    tokenizer = RetokenizeModifier(
+      src=dict(detokenize='spaces', tokenize=f'spaces'),
+      trg=dict(detokenize='spaces', tokenize=f'spaces'))
+
+    test_pair = [
+      'This is a sentence split by spaces .',
+      'Is this a sentence that is split by spaces ?',
+      '0-1 1-0 2-2 3-3 4-6 5-7 6-8 7-9',
+    ]
+
+    self.assertEqual(tokenizer('\t'.join(test_pair)).split('\t'), test_pair)
+
+  def test_multiple_spaces(self):
+    """Spaces tokenizer should have no impact on the output, but it will 
+    normalize spaces due to how it is implemented."""
```

Comment:
I suspect we do not want to normalize spaces when some of the modifiers specifically augment them.

Something like `[m.groupdict() for m in re.finditer('(?P<space>\s+)|(?P<token>\S+)', s]` would get a string segmented into tokens and spaces.

However, would this only be an issue when going from space-to-space?

Reply:
I've added a test case for it now to document this behaviour. All our modifiers split on one-or-more spaces. I'd like to change that, but think something like #29 is the better approach, as opposed to trying to getting this right in each modifier implementation.

Reply:
Agreed